File: solutions/02 photosearcher/java/src/eu/sioux/tdd/photosearcher/FuzzyDate.java
package eu.sioux.tdd.photosearcher;

import java.util.Date;

public class FuzzyDate {
    public Date startDate;
    public Date endDate;

    public enum Quarters {
        Q1,
        Q2,
        Q3,
        Q4
    }

    public enum YearType {
        Start,
        Halfway,
        End
    }

    public FuzzyDate() {
        startDate = new Date(0);
        endDate = new Date(0);
    }

    public FuzzyDate(String string) {
        if (string == "now") {
            startDate = new Date();
            endDate = new Date();
        }
    }

    public FuzzyDate(Quarters quart, int year) {
        switch (quart) {
            case Q1:
                startDate = new Date(ConvertYear(year), 0, 1);
                endDate = new Date(ConvertYear(year), 2, 31);
                break;
            case Q2:
                startDate = new Date(ConvertYear(year), 3, 1);
                endDate = new Date(ConvertYear(year), 5, 30);
                break;
            case Q3:
                startDate = new Date(ConvertYear(year), 6, 1);
                endDate = new Date(ConvertYear(year), 8, 30);
                break;
            case Q4:
                startDate = new Date(ConvertYear(year), 9, 1);
                endDate = new Date(ConvertYear(year), 11, 31);
                break;
        }
    }

    public FuzzyDate(YearType type, int year) {
        switch (type) {
            case Start:
                startDate = new Date(ConvertYear(year), 0, 1);
                endDate = new Date(ConvertYear(year), 3, 30);
                break;
            case Halfway:
                startDate = new Date(ConvertYear(year), 4, 1);
                endDate = new Date(ConvertYear(year), 7, 31);
                break;
            case End:
                startDate = new Date(ConvertYear(year), 8, 1);
                endDate = new Date(ConvertYear(year), 11, 31);
                break;
        }
    }

    private int ConvertYear(int year) {
        return (year - 1900);
    }

    public Date GetFuzzyStart() {
        return startDate;
    }

    public Date GetFuzzyEnd() {
        return endDate;
    }

    public Date GetFuzzyRange() {
        return new Date(endDate.getTime() - startDate.getTime());
    }

    public boolean contains(Date lookUpDate) {
        boolean startContains = startDate.before(lookUpDate) || startDate.equals(lookUpDate);
        boolean endContains = endDate.after(lookUpDate) || endDate.equals(lookUpDate);
        return (startContains && endContains);
    }
}

File: exercises/02 photosearcher/cs/PhotoSearcher/FuzzyDate.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace PhotoSearcher
{
    public class FuzzyDate
    {
        private DateTime startDate;
        private DateTime endDate;

        public enum Quarters
        {
            Q1,
            Q2,
            Q3,
            Q4
        }

        public enum YearType
        {
            Start,
            Halfway,
            End
        }

        public FuzzyDate()
        {
            startDate = new DateTime(0);
            endDate = new DateTime(0);
        }

        public FuzzyDate(string str)
        {
            if (str == "now")
            {
                startDate = DateTime.Now;
                endDate = DateTime.Now;
            }
        }

        public FuzzyDate(Quarters quart, int year)
        {
            switch (quart)
            {
                case Quarters.Q1:
                    startDate = new DateTime(year, 1, 1);
                    endDate = new DateTime(year, 3, 31);
                    break;
                case Quarters.Q2:
                    startDate = new DateTime(year, 4, 1);
                    endDate = new DateTime(year, 6, 30);
                    break;
                case Quarters.Q3:
                    startDate = new DateTime(year, 7, 1);
                    endDate = new DateTime(year, 9, 30);
                    break;
                case Quarters.Q4:
                    startDate = new DateTime(year, 10, 1);
                    endDate = new DateTime(year, 12, 31);
                    break;
            }
        }

        public FuzzyDate(YearType type, int year)
        {
            switch (type)
            {
                case YearType.Start:
                    startDate = new DateTime(year, 1, 1);
                    endDate = new DateTime(year, 4, 30);
                    break;
                case YearType.Halfway:
                    startDate = new DateTime(year, 5, 1);
                    endDate = new DateTime(year, 8, 31);
                    break;
                case YearType.End:
                    startDate = new DateTime(year, 9, 1);
                    endDate = new DateTime(year, 12, 31);
                    break;
            }
        }

        public DateTime GetFuzzyStart()
        {
            return startDate;
        }

        public DateTime GetFuzzyEnd()
        {
            return endDate;
        }

        public DateTime GetFuzzyRange()
        {
            return new DateTime(endDate.Ticks - startDate.Ticks);
        }

        public bool Contains(DateTime lookUpDate)
        {
            bool startContains = startDate.CompareTo(lookUpDate) <= 0;
            bool endContains = endDate.CompareTo(lookUpDate) >= 0;
            return (startContains && endContains);
        }
    }
}
